### ISSUE TITLE
markup (HTML or XML) should be stripped before generating the section id

### ISSUE THREAD

**Issue by mojavelinux** [asciidoctor/asciidoctor#794]
The HTML or XML markup should be stripped from the title before the section id is generated. Otherwise, the id generated is nonsensical.

**Comment by mojavelinux** [comment]
This is related to #840.

**Comment by mojavelinux** [comment]
This will be resolved when #840 is resolved.

**Comment by mojavelinux** [comment]
Here's an example:

```
== https://www.ecosia.org/search?q=asciidoctor[Asciidoctor]
```

which produces the ID:

```
_a_href_https_www_ecosia_org_search_q_asciidoctor_asciidoctor_a
```

**Comment by mojavelinux** [comment]
I'd like to take this a bit further. When generating an ID from section title, all of the following patterns should be replaced with blank:

* XML tags (opening tag including attributes and closing tag)
* character references
* non-word, non-space characters

Once that is done, all space characters should be replaced with the idseparator char.

Not only does this make more sense, it's also consistent with how IDs are generated on GitHub (which we cannot control).

Based on this proposal, here's an example of IDs that would be generated by section titles by default:

* `https://www.ecosia.org/search?q=asciidoctor[Asciidoctor]` ⇒ `_asciidoctor`
* `What's New?` ⇒ `_whats_new` (notice no separator before the "s")
* `Download & Install` ⇒ `_download_install`

(The same rules for idprefix and idseparator would still apply).

This will be a breaking change, though that's sort of implied by what is being requested in this issue.

**Comment by elextr** [comment]
> _a_href_https_www_ecosia_org_search_q_asciidoctor_asciidoctor_a

Well its an id :)

But not one a human would reasonably reference.

But what would you generate in its place?

**Comment by mojavelinux** [comment]
See last comment.

**Comment by elextr** [comment]
yeah, you just beat me :)

Whats the tag generated from? There are two things in the source that say `asciidoctor`?

By applying your rules to the source I still get the long one, or do you apply them to the output after the link has become a tag (and what about non-HTML output?)  what about:

```
== _foo_func()_
```

Since thats in a tag created by the quote markup won't it just be blank?

**Comment by elextr** [comment]
I agree it isn't optimal as it stands, and drives users to explicit ids, but my use-case is when documenting APIs I have lots of headings like
```
==== Function `foo_func()`

==== Function `bar_func()`
```
and at the moment it includes `foo_func` and `bar_func` in the ids (albiet with `code` twice, which as you point out is rubbish) so they are different.  If the contents of tags are dropped then they will both be `_function`.

**Comment by elextr** [comment]
I'm sort of thinking that the ids should be generated from the source, that makes it more accessible to the writer, and also gets around things like attributes substituted in the section titles.

```
== Blah{attr}  -> _blahattr

== Function `foo_func()`  -> _function_foo_func
```

**Comment by mojavelinux** [comment]
The purpose of this issue isn't to change the ID generation strategy, just to address a flaw in it. There's a separate issue where I propose making the ID generator an extension point. I believe that is the extension point you'd use if you wanted to generate IDs in a different way. See #576.
